roofit/roofit/src/RooIntegralMorph.cxx

```diff
@@ -422,8 +421,8 @@ void RooIntegralMorph::MorphCacheElem::calculate(TIterator* dIter)
     _rf1->findRoot(x1,x1,xMax,y) ;
     _rf2->findRoot(x2,x2,xMax,y) ;
 
-    _x->setVal(x1) ; double f1x1 = _pdf1->getVal(&nsetTmp) ;
```

Comment:
For my curiosity, why this was problematic ? 
Is the passed ArgSet pointer kept inside the class ?

Reply:
It's a long story actually. For a RooAbsPdf, the last norm sets are cached in two places:
* the `RooAbsPdf::_normSet` member
* the `RooAbsProxy::_nset` member for all proxies.

With the other PRs I made, there is now perfect protection against an invalid `_normSet` of the PDF because we can always reset it if the pdf if evaluated with another norm set.

For the proxies, we can't do that, because there are cases where the proxies need to explicitly use the last normalization set: for the numeric integration of a PDF, the integral calls `getVal(nullptr)` to disable the normalization of the top-level PDF, but it assumes all servers are still correctly normalized themselves via a previous call to `getVal()` with a norm set. Hence we are not allowed to reset the proxy `_nset`s.

Since we are not allowed to reset the proxy normsets, if this `_pdf1` is evaluated with `getVal()` by a RooAbsIntegral after this line, you access invalid stack memory, as `nsetTmp` was on the stack,

That's at least how I understand it now after long investigations.

Reply:
I think this is exactly why the `function.getVal(_funcNormSet)` call happens in RooRealIntegral: to set the proxy `_nsets` to a valid normalization set. Only that it has no effect if `_funcNormSet` is a `nullptr`, which is always the case except for pass-though integrals, for which this would not be needed to begin with.

Reply:
I understand it now. 
The problem is if one would store the normset by value instead of pointer, it will be not efficient.

Reply:
Thank you very much for the long explanation !!